File: GenerateQRView/src/main/java/daniel/perez/generateqrview/GenerateQR.java
package daniel.perez.generateqrview;

import android.annotation.SuppressLint;
import android.content.Intent;
import android.graphics.Bitmap;
import android.net.Uri;
import android.os.Bundle;
import android.os.ParcelFileDescriptor;
import android.util.Log;

import androidx.annotation.Nullable;
import androidx.lifecycle.ViewModelProvider;
import androidx.lifecycle.ViewModelProviders;

import java.io.File;
import java.io.FileNotFoundException;
import java.io.FileOutputStream;
import java.io.IOException;
import java.util.Collections;
import java.util.Objects;

import javax.inject.Inject;

import daniel.perez.core.BaseActivity;
import daniel.perez.core.DialogClosable;
import daniel.perez.core.DialogStarter;
import daniel.perez.core.Utils;
import daniel.perez.generateqrview.databinding.ActivityGenerateQrBinding;
import daniel.perez.generateqrview.di.GenerateQRComponentProvider;
import daniel.perez.ical.EventSpec;
import daniel.perez.ical.ICalSpec;
import daniel.perez.ical.Parser;
import io.reactivex.functions.Consumer;
import okio.BufferedSink;
import okio.Okio;
import daniel.perez.core.service.FileManager;
import daniel.perez.core.service.qr.QrAttempt;
import daniel.perez.core.service.qr.QrBitmapGenerator;
import daniel.perez.core.model.ViewCalendarData;
import daniel.perez.core.model.ViewEvent;

public class GenerateQR extends BaseActivity implements Consumer<QrAttempt>, DialogClosable
{
    private GenerateQrViewModel viewModel;
    private ActivityGenerateQrBinding binding;
    private ICalSpec currentEvent;
    @Inject ViewModelProvider.Factory viewModelFactory;
    @Inject QrBitmapGenerator generator;
    @Inject FileManager fileManager;
    @Inject DialogStarter dialogStarter;

    @SuppressLint("CheckResult")
    @Override
    protected void onCreate(Bundle savedInstanceState) {
        ((GenerateQRComponentProvider) getApplication())
                .provideGenerateQRComponent()
                .inject(this);
        super.onCreate(savedInstanceState);
        binding = ActivityGenerateQrBinding.inflate(getLayoutInflater());
        setContentView(binding.getRoot());
        viewModel = ViewModelProviders.of(this, viewModelFactory).get(GenerateQrViewModel.class);

        viewModel.qrImages()
                .subscribe(this);

        if(getIntent() != null && getIntent().getStringExtra("FILE_PATH") != null)
        {
            File file = new File(getIntent().getStringExtra("FILE_PATH"));
            ICalSpec ical = Parser.INSTANCE.parse(file);
            currentEvent = ical;
            Log.i("ICALFILEPATH", ical.getEvents().get(0).toString());
            fillInFields();
        }

        binding.startDateEditText.setOnClickListener(view -> startDateDialog());
        binding.startTimeEditText.setOnClickListener(view -> startTimeDialog());
        binding.endDateEditText.setOnClickListener(view -> endDateDialog());
        binding.endTimeEditText.setOnClickListener(view -> endTimeDialog());
        binding.addBtn.setOnClickListener(view -> generateQR());
    }

    //fill in editText and other fields when event is being edited
    public void fillInFields(){
        EventSpec event = currentEvent.getEvents().get(0);
        binding.titleEditText.setText(event.getTitle());
        binding.startDateTextView.setText(event.getStartDate());
        binding.startTimeTextView.setText(event.getStartTime());
        binding.endDateTextView.setText(event.getEndFormatted("MM-dd-yyyy"));
        binding.endTimeTextView.setText(event.getEndFormatted("hh:mm a"));
        binding.locationEditText.setText(event.getLocation());
        binding.descriptionEditText.setText(event.getDescription());
    }

    @Override
    public void accept(QrAttempt attempt) {
        if (attempt instanceof QrAttempt.Success) {
            Bitmap bitmap = ((QrAttempt.Success) attempt).getBitmap();
            //binding.QRImage.setImageBitmap(bitmap);
        } else if (attempt instanceof QrAttempt.Failure) {
            Throwable e = ((QrAttempt.Failure) attempt).getE();
            e.printStackTrace();
            Utils.toastShort(getApplicationContext(), Objects.requireNonNull(e.getLocalizedMessage()));
        }
    }

    public void generateQR() {
        // Handle the dates
        int[] startMonthDayYear = Utils.toInts(binding.startDateTextView.getText().toString().split("/"));
        int[] endMonthDayYear = Utils.toInts(binding.endDateTextView.getText().toString().split("/"));

        // Handle the hours and minutes
        int[] startHourMinutes = Utils.toInts(binding.startTimeTextView.getText().toString().split(":"));
        int[] endHourMinutes = Utils.toInts(binding.endTimeTextView.getText().toString().split(":"));

        EventSpec event = new EventSpec.Builder(0)
                .title(binding.titleEditText.getText().toString())
                .description(binding.descriptionEditText.getText().toString())
                .location(binding.locationEditText.getText().toString())
                .start(Utils.toZonedDateTime(startMonthDayYear, startHourMinutes))
                .end(Utils.toZonedDateTime(endMonthDayYear, endHourMinutes))
                .build();

        currentEvent = new ICalSpec.Builder()
                .fileName(binding.titleEditText.getText().toString())
                .addEvent(event)
                .build();

        // Write the file to the file system
        viewModel.saveFile(currentEvent);

        ViewEvent viewEvent = new ViewEvent(event.getTitle(),
                event.getDescription(),
                event.getStartTime(),
                event.getStartDate(),
                event.getStartInstantEpoch(),
                event.getEndEpochMilli(),
                event.getLocation(),
                event.text(),
                generator.forceGenerate(event.text())
        );
        ViewCalendarData calendar = new ViewCalendarData(currentEvent.getFileName(), Collections.singletonList(viewEvent));
        ((DialogStarter) getApplication()).startQrDialog(this, calendar);
    }

    // Handle the user choosing a place to store the file
    @Override
    protected void onActivityResult(int requestCode, int resultCode, @Nullable Intent data) {
        super.onActivityResult(requestCode, resultCode, data);
        if (requestCode == 1) {
            Uri uri;
            if (data != null) {
                uri = data.getData();
                try {
                    ParcelFileDescriptor pfd = getContentResolver().openFileDescriptor(uri, "w");

                    FileOutputStream fileOutputStream = new FileOutputStream(pfd.getFileDescriptor());
                    try (BufferedSink sink = Okio.buffer(Okio.sink(fileOutputStream))) {
                        sink.writeUtf8(currentEvent.text());
                    } catch (IOException e) {
                        Log.i("FILEWRITER", e.getMessage());
                        e.printStackTrace();
                    }
                    pfd.close();
                } catch (FileNotFoundException e) {
                    Log.i("FILEWRITER", e.getMessage());
                    e.printStackTrace();
                } catch (IOException e) {
                    e.printStackTrace();
                }
                Log.i("FILEWRITER", "Wrote File");
            }
        }
    }

    public void startDateDialog()
    {
        dialogStarter.startDateDialog(this, binding.startDateTextView);
    }

    public void startTimeDialog()
    {
        dialogStarter.startTimeDialog(this, binding.startTimeTextView);
    }

    public void endDateDialog()
    {
        dialogStarter.startDateDialog(this, binding.endDateTextView);
    }

    public void endTimeDialog()
    {
        dialogStarter.startTimeDialog(this, binding.endTimeTextView);
    }


    @Override
    public void close()
    {
        finish();
    }
}

File: GenerateQRView/src/main/java/daniel/perez/generateqrview/GenerateQR.kt
package daniel.perez.generateqrview

import android.annotation.SuppressLint
import android.content.Intent
import android.os.Bundle
import android.view.View
import android.widget.AdapterView
import android.widget.ArrayAdapter
import androidx.lifecycle.ViewModelProvider
import androidx.lifecycle.ViewModelProviders
import com.jakewharton.rxbinding4.view.clicks
import daniel.perez.core.*
import daniel.perez.core.db.UiNewEvent
import daniel.perez.core.db.toViewEvent
import daniel.perez.core.model.DateSetEvent
import daniel.perez.core.model.Half
import daniel.perez.core.model.TimeSetEvent
import daniel.perez.core.model.ViewEvent
import daniel.perez.generateqrview.databinding.ActivityGenerateQrBinding
import daniel.perez.generateqrview.di.GenerateQRComponentProvider
import daniel.perez.ical.Parser.parse
import io.reactivex.rxjava3.android.schedulers.AndroidSchedulers
import timber.log.Timber
import java.io.File
import java.time.LocalDateTime
import javax.inject.Inject

class GenerateQR : BaseActivity(), DialogClosable, AdapterView.OnItemSelectedListener
{
    private lateinit var viewModel: GenerateQrViewModel
    private lateinit var binding: ActivityGenerateQrBinding
    private lateinit var currentEvent: ViewEvent
    private var startTime = TimeSetEvent(0, 0, Half.AM)
    private var endTime = TimeSetEvent(0, 0, Half.AM)
    private var startDate = DateSetEvent(1, 31, 1999)
    private var endDate = DateSetEvent(1, 31, 1999)
    @Inject lateinit var viewModelFactory: ViewModelProvider.Factory
    @Inject lateinit var dialogStarter: DialogStarter
    @Inject lateinit var activityResultActions: ActivityResultActions

    @SuppressLint("CheckResult")
    override fun onCreate(savedInstanceState: Bundle?)
    {
        (application as GenerateQRComponentProvider)
                .provideGenerateQRComponent()
                .inject(this)
        super.onCreate(savedInstanceState)
        binding = ActivityGenerateQrBinding.inflate(layoutInflater)
        setContentView(binding.root)
        viewModel = ViewModelProviders.of(this, viewModelFactory).get(GenerateQrViewModel::class.java)

        if (intent != null && intent.getStringExtra("FILE_PATH") != null)
        {
            val file = File(intent.getStringExtra("FILE_PATH"))
            val ical = parse(file)
//            currentEvent = ical
            Timber.i(ical.events[0].toString())
            fillInFields()
        }

        ArrayAdapter.createFromResource(this, R.array.event_recurrence_options_list, R.layout.spinner_text)
                .also { adapter ->
                    adapter.setDropDownViewResource(android.R.layout.simple_spinner_dropdown_item)
                    binding.recurrenceSpinner.adapter = adapter}
        binding.recurrenceSpinner.onItemSelectedListener = this
        disposables += binding.startDateEditText.clicks().subscribe { startDateDialog() }
        binding.startTimeEditText.clicks().subscribe { startTimeDialog() }
        binding.endDateEditText.clicks().subscribe { endDateDialog() }
        binding.endTimeEditText.clicks().subscribe { endTimeDialog() }
        binding.addBtn.clicks().subscribe { saveEvent() }
    }



    //fill in editText and other fields when event is being edited
    private fun fillInFields()
    {
        val event = currentEvent

        //TODO("Implement the enum properly, calculate if its morning or after noon for AM, PM")
        startTime = TimeSetEvent(event.startDateTime.hour, event.startDateTime.minute, Half.AM)
        endTime = TimeSetEvent(event.endDateTime.hour, event.endDateTime.minute, Half.PM)
        startDate = DateSetEvent(event.startDateTime.monthValue, event.startDateTime.dayOfMonth, event.startDateTime.year)
        endDate = DateSetEvent(event.endDateTime.monthValue, event.endDateTime.dayOfMonth, event.endDateTime.year)

        binding.titleEditText.setText(event.title)
        binding.startDateTextView.text = event.startDatePretty()
        binding.startTimeTextView.text = event.startTimePretty()
        binding.endDateTextView.text = event.endDatePretty()
        binding.endTimeTextView.text = event.endTimePretty()
        binding.locationEditText.setText(event.location)
        binding.descriptionEditText.setText(event.description)
    }

    private fun areValidFields() : Boolean {
        if (binding.titleEditText.text.isBlank()) {
            applicationContext.snackbarShort(binding.titleEditText, getString(R.string.enter_valid_title))
            return false
        }

        if (binding.startDateTextView.text.toString().equals(getString(R.string.select_date))) {
            applicationContext.snackbarShort(binding.startDateTextView, getString(R.string.enter_valid_start_date))
            return false
        }

        //if end date is empty, set it to be the same as start date
        if (binding.endDateTextView.text.toString().equals(getString(R.string.select_date))) {
            endDate = startDate
        }

        return true
    }

    private fun saveEvent()
    {
        if (!areValidFields()) {
            return
        }
        val uiNewEvent: UiNewEvent = extractEventFromUi()
        disposables += viewModel.saveEvent(uiNewEvent)
                .observeOn( AndroidSchedulers.mainThread() )
                .subscribe {
                    currentEvent = it.toViewEvent()
                    dialogStarter.startQrDialog(this, it.toViewEvent())
                }
    }

    private fun extractEventFromUi(): UiNewEvent
    {
        Timber.i("Start_Date: ${startDate.string()}")
        Timber.i("Start_Time: $startTime")
        Timber.i("End_Date: ${endDate.string()}")
        Timber.i("End_Time: $endTime")
        val uiNewEvent: UiNewEvent
        with(binding)
        {
            uiNewEvent = UiNewEvent(
                    titleEditText.text.toString(),
                    descriptionEditText.text.toString(),
                    locationEditText.text.toString(),
                    LocalDateTime.parse("${startDate.string()}T${startTime.string()}" ), //todo add method into Utils.kt
                    LocalDateTime.parse("${endDate.string()}T${endTime.string()}"),
            )
        }

        return uiNewEvent
    }

    // Handle the user choosing a place to store the file
    override fun onActivityResult(requestCode: Int, resultCode: Int, data: Intent?)
    {
        super.onActivityResult(requestCode, resultCode, data)
        if (requestCode == RequestCodes.REQUEST_SAF)
        {
            if (data != null)
            {
                disposables += activityResultActions.actionCreateDocumentEvent(this, currentEvent, data)
                        .subscribe {
                            when(it)
                            {
                                ActionResult.Success -> toastShort("Wrote File Successfully")
                                is ActionResult.Failure ->
                                {
                                    Timber.e(it.t)
                                    toastShort("Error writing file")
                                }
                                ActionResult.InTransit ->
                                {
                                    Timber.i("Writing file...")
                                }
                            }
                        }
            }
        }
    }

    private fun startDateDialog()
    {
        disposables.add(dialogStarter.startDateDialog(this)
                .subscribe {
                    startDate = it
                    binding.startDateTextView.text = "${it.month}-${it.dayOfMonth}-${it.year}"
                })
    }

    private fun endDateDialog()
    {
        disposables.add( dialogStarter.startDateDialog( this )
                .subscribe {
                    endDate = it
                    binding.endDateTextView.text = "${it.month}-${it.dayOfMonth}-${it.year}"

                } )
    }

    private fun startTimeDialog()
    {
        disposables.add(dialogStarter.startTimeDialog(this)
                .subscribe {
                    startTime = it
                    binding.startTimeTextView.text = getAMPMTimeFormat("${it.hour} ${it.minute}")
                } )
    }

    private fun endTimeDialog()
    {
        disposables.add(dialogStarter.startTimeDialog(this)
                .subscribe {
                    endTime = it
                    binding.endTimeTextView.text = getAMPMTimeFormat("${it.hour} ${it.minute}")
                })
    }

    override fun onDestroy() {
        super.onDestroy()
        disposables.dispose()
    }

    override fun close()
    {
        finish()
    }

    override fun onItemSelected(parent: AdapterView<*>?, view: View?, position: Int, id: Long) {

    }

    override fun onNothingSelected(p0: AdapterView<*>?) {

    }
}